### ISSUE TITLE
Feature Request: Customized 404 pages (without changing code and recompiling)

### ISSUE THREAD

**Issue by mojoTX** [reyk/httpd#62]
I would like the ability to have more control over the system-generated error pages, besides modifying C code myself and recompiling.

In other words, I don't want my "404 Not Found" pages saying, "OpenBSD httpd" at the bottom and using the Comic Sans font. 

The nginx approach using the "error_page" keyword to specify a URI for a given HTTP code would be relatively straightforward to implement, I think. If there is nothing in the httpd.conf file for that server and a given return code, then the fall-back could still use the hard-coded version.

**Comment by tvlooy** [comment]
I'm not going to send this as a patch myself because I don't need it but as an exercise I added such an option https://gist.github.com/tvlooy/8f5c9253f9e283eaee9a77e9721daa52

**Comment by vetelko** [comment]
Love spartan features and behavior of the httpd, but thanks for this request. 404 is IMO the only one error page which **have to** "communicate" with the user, provide him with, at least, a link to the site home page. Otherwise we lost this user in most cases.

**Comment by afh** [comment]
@reyk What are your thoughts on the [changes](https://gist.github.com/tvlooy/8f5c9253f9e283eaee9a77e9721daa52) that @tvlooy did as an exercise  to add support for custom error pages?

IMO it would make a useful addition to `httpd (8)` and I'd excited to see it in one the the next OpenBSD releases.

**Comment by jaywilliams** [comment]
I throughly agree. I love the minimalist featureset of httpd, but having a fixed 404 error page is not very visitor friendly.

**Comment by tvlooy** [comment]
what I would like is have an option to handle requests with factcgi if the requested location is not a file or directory. Like in Apache:

```
RewriteCond %{REQUEST_FILENAME} !-d
RewriteCond %{REQUEST_FILENAME} !-f
RewriteRule ^ index.php [QSA,L]
```

**Comment by vetelko** [comment]
@tvlooy 
... or try_files in nginx

**Comment by B4rb3rouss** [comment]
@vetelko said everything. This is a small improvement very useful for visitor. Hope to see this in httpd too.

**Comment by ghost** [comment]
I agree, this would be very useful. I have a client's new site running on httpd but there are links out there in the wild to pages from their old site that no longer exist. I'd love to be able to give visitors a custom 404 page with a link to the current homepage if they happen to click on them.

**Comment by vetelko** [comment]
@jsfierro in this case you should try to redirect (if old url structure allows this) user using 301 code from old url to new url or to homepage

**Comment by ghost** [comment]
@vetelko 

That's that what I'm doing now, but it's kind of tricky. The old site had _tons_ of pages, with lots of duplicate content, non-obvious subdirectory names, etc. I have many of the important ones redirected, but there's a lot I am missing, so a custom 404 to catch those would be great.

**Comment by vetelko** [comment]
if som logical pattern in old url strcuture is missing than you are right. can't you switch to nginx for now? I doubt custom 404 via directive in httpd.conf will be solved anytime soon :(

**Comment by Ranguvar** [comment]
Just want to note that @tvlooy's patch is working brilliantly for me as of 6.3 httpd.

Thank you! Hopefully this is looked at for a future release.

**Comment by Ranguvar** [comment]
Updated for 6.4: https://gist.github.com/Ranguvar/7fdec9c8f0a55edad4cdabb4c5c59f79
Not working for me on 6.4, may be patch issue or something else.
